File: typescript/src/proxy.ts
import { spawn, ChildProcess } from 'child_process';
import * as path from 'path';
import * as fs from 'fs';
import * as net from 'net';
import { ensureBinary, getFullBinaryPath } from './binary';
import type { ProxyMode, RecordingOptions, PlaybackOptions, Inventory } from './types';

/**
 * Find an available port by binding to port 0 and releasing it
 */
async function getAvailablePort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const server = net.createServer();
    server.listen(0, '127.0.0.1', () => {
      const address = server.address();
      if (address && typeof address === 'object') {
        const port = address.port;
        server.close(() => resolve(port));
      } else {
        server.close(() => reject(new Error('Failed to get port from server address')));
      }
    });
    server.on('error', reject);
  });
}

/**
 * Wait for a port to become available (accepting connections)
 * Uses TCP connection attempts with exponential backoff
 */
async function waitForPort(port: number, timeoutMs: number = 60000): Promise<void> {
  const startTime = Date.now();
  let delay = 50; // Start with 50ms delay

  while (Date.now() - startTime < timeoutMs) {
    try {
      await new Promise<void>((resolve, reject) => {
        const socket = net.createConnection({ port, host: '127.0.0.1' });
        socket.on('connect', () => {
          socket.destroy();
          resolve();
        });
        socket.on('error', reject);
        socket.setTimeout(1000, () => {
          socket.destroy();
          reject(new Error('Connection timeout'));
        });
      });
      return; // Port is open
    } catch {
      // Port not ready yet, wait and retry
      await new Promise((resolve) => setTimeout(resolve, delay));
      delay = Math.min(delay * 1.5, 500); // Exponential backoff, max 500ms
    }
  }

  throw new Error(`Timeout waiting for port ${port} to become available`);
}

/**
 * Represents a running proxy instance
 */
export class Proxy {
  public readonly mode: ProxyMode;
  public readonly inventoryDir: string;
  public readonly entryUrl?: string;
  public readonly deviceType?: string;

  private _port: number;
  private process?: ChildProcess;

  constructor(
    mode: ProxyMode,
    port: number,
    inventoryDir: string,
    entryUrl?: string,
    deviceType?: string
  ) {
    this.mode = mode;
    this._port = port;
    this.inventoryDir = inventoryDir;
    this.entryUrl = entryUrl;
    this.deviceType = deviceType;
  }

  /**
   * Get the actual port number
   */
  get port(): number {
    return this._port;
  }

  /**
   * Set the child process
   */
  setProcess(proc: ChildProcess): void {
    this.process = proc;
  }

  /**
   * Stop the proxy gracefully
   * Sends SIGTERM signal (cross-platform)
   */
  async stop(): Promise<void> {
    if (!this.process) {
      throw new Error('Proxy is not running');
    }

    return new Promise((resolve, reject) => {
      if (!this.process) {
        reject(new Error('Proxy is not running'));
        return;
      }

      // Set up timeout for forceful termination
      const timeout = setTimeout(() => {
        if (this.process) {
          this.process.kill('SIGKILL');
          reject(new Error('Proxy did not stop gracefully, killed forcefully'));
        }
      }, 10000);

      // Listen for exit
      this.process.once('exit', (code, signal) => {
        clearTimeout(timeout);
        // Accept clean exits: code 0, SIGTERM, SIGINT
        if (code === 0 || code === null || signal === 'SIGTERM' || signal === 'SIGINT') {
          resolve();
        } else {
          reject(new Error(`Proxy exited with code ${code} signal ${signal}`));
        }
      });

      // Send platform-appropriate signal
      try {
        if (process.platform === 'win32') {
          // On Windows, use the signal subcommand to send CTRL_BREAK
          const binaryPath = getFullBinaryPath();
          const { spawnSync } = require('child_process');
          const result = spawnSync(
            binaryPath,
            ['signal', '--pid', this.process.pid!.toString(), '--kind', 'ctrl-break'],
            { stdio: 'pipe' }
          );

          if (result.error) {
            clearTimeout(timeout);
            reject(new Error(`Failed to send signal: ${result.error.message}`));
            return;
          }

          if (result.status !== 0) {
            clearTimeout(timeout);
            const stderr = result.stderr?.toString() || '';
            reject(new Error(`Signal command failed with exit code ${result.status}: ${stderr}`));
            return;
          }
        } else {
          // On Unix, use standard SIGTERM
          this.process.kill('SIGTERM');
        }
      } catch (err) {
        clearTimeout(timeout);
        reject(err);
      }
    });
  }

  /**
   * Check if the proxy is still running
   */
  isRunning(): boolean {
    if (!this.process) {
      return false;
    }

    try {
      process.kill(this.process.pid!, 0);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Wait for the proxy to exit
   */
  async wait(): Promise<void> {
    if (!this.process) {
      throw new Error('Proxy is not running');
    }

    return new Promise((resolve, reject) => {
      this.process!.once('exit', (code) => {
        if (code === 0) {
          resolve();
        } else {
          reject(new Error(`Proxy exited with code ${code}`));
        }
      });
    });
  }

  /**
   * Load the inventory for this proxy
   */
  async getInventory(): Promise<Inventory> {
    const inventoryPath = path.join(this.inventoryDir, 'index.json');
    return loadInventory(inventoryPath);
  }
}

/**
 * Start a recording proxy
 */
export async function startRecording(options: RecordingOptions): Promise<Proxy> {
  await ensureBinary();

  const binaryPath = getFullBinaryPath();

  // Get an available port if not specified or if 0
  let port: number;
  if (options.port === undefined || options.port === 0) {
    port = await getAvailablePort();
  } else {
    port = options.port;
  }

  const deviceType = options.deviceType || 'mobile';
  const inventoryDir = options.inventoryDir || './inventory';

  // Build command
  const args: string[] = ['recording'];

  // Add entry URL if provided
  if (options.entryUrl) {
    args.push(options.entryUrl);
  }

  // Always specify the port explicitly
  args.push('--port', port.toString());

  // Add device type
  args.push('--device', deviceType);

  // Add inventory directory
  args.push('--inventory', inventoryDir);

  // Add extra URLs
  if (options.extraUrls) {
    for (const url of options.extraUrls) {
      args.push('--extra-url', url);
    }
  }

  // Start the process
  const spawnOptions: any = {
    stdio: ['ignore', 'inherit', 'inherit'],
    detached: false,
  };

  if (process.platform === 'win32') {
    spawnOptions.windowsVerbatimArguments = false;
  }

  const proc = spawn(binaryPath, args, spawnOptions);

  const proxy = new Proxy('recording', port, inventoryDir, options.entryUrl, deviceType);
  proxy.setProcess(proc);

  // Handle early exit
  let exited = false;
  proc.on('exit', (code) => {
    exited = true;
    if (code !== 0 && code !== null) {
      console.error(`Proxy process exited early with code ${code}`);
    }
  });

  // Wait for the port to become available
  try {
    await waitForPort(port, 15000);
  } catch (err) {
    if (exited) {
      throw new Error('Proxy process exited before port became available');
    }
    throw err;
  }

  return proxy;
}

/**
 * Start a playback proxy
 */
export async function startPlayback(options: PlaybackOptions): Promise<Proxy> {
  await ensureBinary();

  const binaryPath = getFullBinaryPath();
  const inventoryDir = options.inventoryDir || './inventory';

  // Verify inventory exists
  const inventoryPath = getInventoryPath(inventoryDir);
  if (!fs.existsSync(inventoryPath)) {
    throw new Error(`Inventory file not found at ${inventoryPath}`);
  }

  // Get an available port if not specified or if 0
  let port: number;
  if (options.port === undefined || options.port === 0) {
    port = await getAvailablePort();
  } else {
    port = options.port;
  }

  // Build command
  const args: string[] = ['playback'];

  // Always specify the port explicitly
  args.push('--port', port.toString());

  // Add inventory directory
  args.push('--inventory', inventoryDir);

  // Add full throttle flag
  if (options.fullThrottle) {
    args.push('--full-throttle');
  }

  // Add passthrough flag
  if (options.passthrough) {
    args.push('--passthrough');
  }

  // Start the process
  const spawnOptions: any = {
    stdio: ['ignore', 'inherit', 'inherit'],
    detached: false,
  };

  if (process.platform === 'win32') {
    spawnOptions.windowsVerbatimArguments = false;
  }

  const proc = spawn(binaryPath, args, spawnOptions);

  const proxy = new Proxy('playback', port, inventoryDir);
  proxy.setProcess(proc);

  // Handle early exit
  let exited = false;
  proc.on('exit', (code) => {
    exited = true;
    if (code !== 0 && code !== null) {
      console.error(`Proxy process exited early with code ${code}`);
    }
  });

  // Wait for the port to become available
  try {
    await waitForPort(port, 15000);
  } catch (err) {
    if (exited) {
      throw new Error('Proxy process exited before port became available');
    }
    throw err;
  }

  return proxy;
}

/**
 * Load an inventory from a JSON file
 */
export async function loadInventory(inventoryPath: string): Promise<Inventory> {
  const data = await fs.promises.readFile(inventoryPath, 'utf8');
  return JSON.parse(data) as Inventory;
}

/**
 * Save an inventory to a JSON file
 */
export async function saveInventory(inventoryPath: string, inventory: Inventory): Promise<void> {
  const data = JSON.stringify(inventory, null, 2);
  await fs.promises.writeFile(inventoryPath, data, 'utf8');
}

/**
 * Get the full path to a resource's content file
 */
export function getResourceContentPath(inventoryDir: string, resource: { contentFilePath?: string }): string {
  if (!resource.contentFilePath) {
    return '';
  }
  return path.join(inventoryDir, resource.contentFilePath);
}

/**
 * Get the path to the index.json file
 */
export function getInventoryPath(inventoryDir: string): string {
  return path.join(inventoryDir, 'index.json');
}

File: golang/proxy.go
package httpplaybackproxy

import (
	"context"
	"fmt"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"sync"
	"time"
)

// ProxyMode represents the mode of the proxy
type ProxyMode string

const (
	ModeRecording ProxyMode = "recording"
	ModePlayback  ProxyMode = "playback"
)

// Proxy represents a running proxy instance
type Proxy struct {
	Mode         ProxyMode
	Port         int
	InventoryDir string
	EntryURL     string     // Only for recording mode
	DeviceType   DeviceType // Only for recording mode
	cmd          *exec.Cmd
	ctx          context.Context
	cancel       context.CancelFunc
	portMutex    sync.RWMutex
}

// RecordingOptions holds options for starting a recording proxy
type RecordingOptions struct {
	EntryURL     string     // Optional: Entry URL to start recording from
	Port         int        // Optional: Port to use (0 = auto-detect)
	DeviceType   DeviceType // Optional: Device type (default: mobile)
	InventoryDir string     // Optional: Inventory directory (default: ./inventory)
}

// PlaybackOptions holds options for starting a playback proxy
type PlaybackOptions struct {
	Port         int
	InventoryDir string
}

// getAvailablePort finds an available port by binding to port 0 and releasing it
func getAvailablePort() (int, error) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return 0, fmt.Errorf("failed to find available port: %w", err)
	}
	port := listener.Addr().(*net.TCPAddr).Port
	listener.Close()
	return port, nil
}

// waitForPort waits for a port to become available (accepting connections)
// Uses TCP connection attempts with exponential backoff
func waitForPort(port int, timeout time.Duration) error {
	startTime := time.Now()
	delay := 50 * time.Millisecond

	for time.Since(startTime) < timeout {
		conn, err := net.DialTimeout("tcp", fmt.Sprintf("127.0.0.1:%d", port), time.Second)
		if err == nil {
			conn.Close()
			return nil // Port is open
		}
		time.Sleep(delay)
		delay = time.Duration(float64(delay) * 1.5)
		if delay > 500*time.Millisecond {
			delay = 500 * time.Millisecond
		}
	}

	return fmt.Errorf("timeout waiting for port %d to become available", port)
}

// StartRecording starts a recording proxy
func StartRecording(opts RecordingOptions) (*Proxy, error) {
	if err := EnsureBinary(); err != nil {
		return nil, fmt.Errorf("failed to ensure binary: %w", err)
	}

	binaryPath, err := GetBinaryPath()
	if err != nil {
		return nil, err
	}

	// Get an available port if not specified
	port := opts.Port
	if port == 0 {
		port, err = getAvailablePort()
		if err != nil {
			return nil, err
		}
	}

	deviceType := opts.DeviceType
	if deviceType == "" {
		deviceType = DeviceTypeMobile
	}

	inventoryDir := opts.InventoryDir
	if inventoryDir == "" {
		inventoryDir = "./inventory"
	}

	// Build command
	ctx, cancel := context.WithCancel(context.Background())
	args := []string{"recording"}

	// Add entry URL if provided
	if opts.EntryURL != "" {
		args = append(args, opts.EntryURL)
	}

	// Always specify the port explicitly
	args = append(args, "--port", strconv.Itoa(port))

	// Add device type
	args = append(args, "--device", string(deviceType))

	// Add inventory directory
	args = append(args, "--inventory", inventoryDir)

	cmd := exec.CommandContext(ctx, binaryPath, args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	setProcAttributes(cmd)

	// Start the process
	if err := cmd.Start(); err != nil {
		cancel()
		return nil, fmt.Errorf("failed to start recording proxy: %w", err)
	}

	proxy := &Proxy{
		Mode:         ModeRecording,
		Port:         port,
		InventoryDir: inventoryDir,
		EntryURL:     opts.EntryURL,
		DeviceType:   deviceType,
		cmd:          cmd,
		ctx:          ctx,
		cancel:       cancel,
	}

	// Wait for the port to become available
	if err := waitForPort(port, 15*time.Second); err != nil {
		// Check if process exited early
		if !proxy.IsRunning() {
			cancel()
			return nil, fmt.Errorf("proxy process exited before port became available")
		}
		cancel()
		return nil, err
	}

	return proxy, nil
}

// StartPlayback starts a playback proxy
func StartPlayback(opts PlaybackOptions) (*Proxy, error) {
	if err := EnsureBinary(); err != nil {
		return nil, fmt.Errorf("failed to ensure binary: %w", err)
	}

	binaryPath, err := GetBinaryPath()
	if err != nil {
		return nil, err
	}

	// Get an available port if not specified
	port := opts.Port
	if port == 0 {
		port, err = getAvailablePort()
		if err != nil {
			return nil, err
		}
	}

	inventoryDir := opts.InventoryDir
	if inventoryDir == "" {
		inventoryDir = "./inventory"
	}

	// Verify inventory exists
	inventoryPath := GetInventoryPath(inventoryDir)
	if _, err := os.Stat(inventoryPath); err != nil {
		return nil, fmt.Errorf("inventory file not found at %s: %w", inventoryPath, err)
	}

	// Build command
	ctx, cancel := context.WithCancel(context.Background())
	args := []string{"playback"}

	// Always specify the port explicitly
	args = append(args, "--port", strconv.Itoa(port))

	args = append(args, "--inventory", inventoryDir)

	cmd := exec.CommandContext(ctx, binaryPath, args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	setProcAttributes(cmd)

	// Start the process
	if err := cmd.Start(); err != nil {
		cancel()
		return nil, fmt.Errorf("failed to start playback proxy: %w", err)
	}

	proxy := &Proxy{
		Mode:         ModePlayback,
		Port:         port,
		InventoryDir: inventoryDir,
		cmd:          cmd,
		ctx:          ctx,
		cancel:       cancel,
	}

	// Wait for the port to become available
	if err := waitForPort(port, 15*time.Second); err != nil {
		// Check if process exited early
		if !proxy.IsRunning() {
			cancel()
			return nil, fmt.Errorf("proxy process exited before port became available")
		}
		cancel()
		return nil, err
	}

	return proxy, nil
}

// Stop stops the proxy gracefully
// Sends SIGTERM (cross-platform)
func (p *Proxy) Stop() error {
	if p.cmd == nil || p.cmd.Process == nil {
		return fmt.Errorf("proxy is not running")
	}

	// Platform-specific process termination (SIGTERM preferred, SIGINT fallback)
	if err := stopProcess(p.cmd.Process); err != nil {
		// If stop fails, cancel the context
		p.cancel()
		return fmt.Errorf("failed to stop process: %w", err)
	}

	return p.waitForExit()
}

// waitForExit waits for the process to exit with proper error handling
func (p *Proxy) waitForExit() error {
	done := make(chan error, 1)
	go func() {
		done <- p.cmd.Wait()
	}()

	select {
	case err := <-done:
		if err != nil {
			// Exit code 130 is expected for SIGINT, -1 for signals, 0 for success
			if exitErr, ok := err.(*exec.ExitError); ok {
				exitCode := exitErr.ExitCode()
				// Windows: 0xc000013a (STATUS_CONTROL_C_EXIT) = 3221225786 or -1073741510
				// Unix: 130 (128 + SIGINT=2) or -1 for signals
				if exitCode == 0 || exitCode == 130 || exitCode == -1 ||
					exitCode == 3221225786 || exitCode == -1073741510 {
					// Normal exit codes for graceful shutdown
					return nil
				}
			}
			// For other signal-related errors, also treat as success
			if err.Error() == "signal: interrupt" {
				return nil
			}
			return fmt.Errorf("proxy exited with error: %w", err)
		}
		// Exit code 0 - success
		return nil
	case <-time.After(10 * time.Second):
		// Force kill if graceful shutdown takes too long
		p.cancel()
		_ = p.cmd.Process.Kill()
		return fmt.Errorf("proxy did not stop gracefully, killed forcefully")
	}
}

// IsRunning checks if the proxy is still running
func (p *Proxy) IsRunning() bool {
	if p.cmd == nil || p.cmd.Process == nil {
		return false
	}

	// Use platform-specific process check (defined in proxy_unix.go and proxy_windows.go)
	return isProcessRunning(p.cmd.Process)
}

// Wait waits for the proxy to exit
func (p *Proxy) Wait() error {
	if p.cmd == nil {
		return fmt.Errorf("proxy is not running")
	}
	return p.cmd.Wait()
}

// GetInventory loads the inventory for this proxy
// This is useful after recording is complete
func (p *Proxy) GetInventory() (*Inventory, error) {
	inventoryPath := filepath.Join(p.InventoryDir, "index.json")
	return LoadInventory(inventoryPath)
}
